### PR TITLE
EarlyStopping add initial warm-up #16793

### PR DESCRIPTION
**Pull request by inonbe**
following the feature requests as described in:
https://github.com/keras-team/keras/issues/16793#issuecomment-1246065854
and after approval I am submitting this PR.
Thanks.

### CODE REVIEW

**Comment by divyashreepathihalli** [comment]
@inonbe can you please add a unit test to callbacks_test.py?

**Comment by inonbe** [comment]
sure, will add unit test to callbacks_test.py

File: keras/callbacks.py [hunk 1]
```diff
@@ -1946,6 +1946,10 @@ class EarlyStopping(Callback):
           of the performance relative to the `baseline`. If no epoch
           improves on `baseline`, training will run for `patience`
           epochs and restore weights from the best epoch in that set.
+      start_from_epoch: Number of initial epochs to wait before starting
```
Comment (fchollet) [inline keras/callbacks.py:1949]: "Number of epochs to wait for before..."

**Review by fchollet — COMMENTED** [review]
Thank you for the PR!

File: keras/callbacks_test.py [hunk 2]
```diff
@@ -1869,6 +1869,46 @@ def set_weight_to_epoch(self, epoch):
...
+            stopper = keras.callbacks.EarlyStopping(
+                monitor="acc",
+                patience=patience,
+                start_from_epoch=start_from_epoch,
+            )
+            hist = model.fit(
+                data, labels, callbacks=[stopper], verbose=0, epochs=20
+            )
+            assert len(hist.epoch) >= patience + start_from_epoch
+
```
Comment (divyashreepathihalli) [inline keras/callbacks_test.py:1899]: add a comment about the difference between the two test cases.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Thank you for the PR! some minor comments.

File: keras/callbacks_test.py [hunk 3]
```diff
@@ -1869,6 +1869,46 @@ def set_weight_to_epoch(self, epoch):
         self.assertEqual(epochs_trained, 5)
         self.assertEqual(early_stop.model.get_weights(), 2)
 
+    def test_EarlyStopping_with_start_from_epoch(self):
+        with self.cached_session():
+            np.random.seed(1337)
+
```
[the remark this answers is not in the document]
Comment (inonbe) [reply keras/callbacks_test.py]: Thanks again for your time and comments - I addressed all comments.

**Comment by divyashreepathihalli** [comment]
@inonbe some tests are failing in `callbacks_test.py`. you can check for more details by clicking on the `details` link above.
here is the failing test - `test_EarlyStopping_final_weights_when_restoring_model_weights`

**Comment by inonbe** [comment]
> @inonbe some tests are failing in `callbacks_test.py`. you can check for more details by clicking on the `details` link below. Also linking it [here](https://source.cloud.google.com/results/invocations/b6ac35b3-dfea-4e84-85fb-381855b73c99/targets). here is the failing test - `test_EarlyStopping_final_weights_when_restoring_model_weights`

sure, sorry for missing this - will fix and update.

**Comment by inonbe** [comment]
to clarify, all tests are passing after last commit.

File: keras/callbacks_test.py [hunk 4]
```diff
@@ -1869,6 +1869,51 @@ def set_weight_to_epoch(self, epoch):
...
+                monitor="acc",
+                patience=patience,
+                start_from_epoch=start_from_epoch,
+            )
+            history = model.fit(
+                data, labels, callbacks=[stopper], verbose=0, epochs=20
+            )
+            # Test 'patience' argument functions correctly when used
+            # in conjunction with 'start_from_epoch'.
+            assert len(history.epoch) >= patience + start_from_epoch
```
[the remark this answers is not in the document]
Comment (inonbe) [reply keras/callbacks_test.py]: Thanks @rchao for your comments.
regarding question 1 - the summation is a one way of checking that 
both existing "patience" mechanism is working as well as the new logic "start_from_epoch".
The "start_from_epoch" was created for the cases were indeed there's no improvement
in the model fit process for at least the amount of initial epochs designated by this "start_from_epoch" 
number. Afterwards, the existing "patience" mechanism should start counting and therefore 
in the worse-case the number of epochs should be at least more than this sum. I hope it makes sense. 
regarding question 2 - of-course I changed the asserts to self.assertX in the commit below.

File: keras/callbacks.py [hunk 5]
```diff
@@ -1946,6 +1946,10 @@ class EarlyStopping(Callback):
           of the performance relative to the `baseline`. If no epoch
           improves on `baseline`, training will run for `patience`
           epochs and restore weights from the best epoch in that set.
+      start_from_epoch: Number of epochs to wait before starting
+          to monitor improvement. This allows a warm-up period in which
```
Comment (fchollet) [inline keras/callbacks.py:1950]: "This allows for"

**Review by fchollet — COMMENTED** [review]
Looking good -- one minor comment

File: keras/callbacks.py (hunk 5 above)
Comment (inonbe) [reply keras/callbacks.py]: Thanks @fchollet - much appreciated! of-course I fixed.

**Review by fchollet — APPROVED** [review]
LGTM, thanks

**Review by divyashreepathihalli — APPROVED** [review]

File: keras/callbacks_test.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rchao) [reply keras/callbacks_test.py]: Thanks @inonbe for the explanation! For question 1, I'm still not fully following: imaging that this training of a model always increases the accuracy as epoch goes on, isn't it true that the early stopping callback will never attempt to stop it, since it's using `monitor="acc"`, and by definition the early stopping callback "Stop training when a monitored metric has stopped improving"? Please let me know if I'm missing anything.

[the remark this answers is not in the document]
Comment (inonbe) [reply keras/callbacks_test.py]: Hi Rick,
you're right when assuming that the accuracy keeps improving, but the purpose of this callback in general
and specially the argument i added are for the cases in which this is not the case.
I.e. when the accuracy is not improving and even degrading initially (regarding the "start_from_epoch")
or degrading after some period of training in the general case of the early stopping callback.
Since this test is relatively short in regard to number of epochs and the input/output data is totally random - 
the accuracy might indeed may randomly degrade occasionally.
I hope its more understood - thanks anyway for your efforts
Inon.
